### PR TITLE
[AKS] Fix Nvidia driver installation on node groups with taints

### PR DESCRIPTION
**Pull request by vrutz**
Similar fix as for EKS, we patch the Nvidia driver with the tolerations derived from taints of the node pools that are GPU-enabled.

### CODE REVIEW

File: python-clusters/create-aks-cluster/cluster.py [hunk 1]
```diff
@@ -10,6 +10,7 @@
```
Comment (amandineslx) [inline python-clusters/create-aks-cluster/cluster.py:10]: missing `CHANGELOG` update

File: python-lib/dku_utils/taints.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
+import json, logging
+from dku_utils.access import _is_none_or_blank
+
+class Taint(dict):
+    def __init__(self, taint):
+        logging.debug("Creating taint from '%s'" % json.dumps(taint))
+        if isinstance(taint, str):
```
Comment (amandineslx) [inline python-lib/dku_utils/taints.py:7]: should we check both `str` and `basestring` for Python2 BW compat?

File: python-lib/dku_kube/nvidia_utils.py [hunk 3]
```diff
@@ -10,11 +14,59 @@ def has_gpu_driver(kube_config_path):
...
+    # Get the Nvidia driver plugin configuration from the repository
+    nvidia_config_raw = requests.get(
+        "https://raw.githubusercontent.com/NVIDIA/k8s-device-plugin/main/deployments/static/nvidia-device-plugin.yml"
+    ).text
+    nvidia_config = yaml.safe_load(nvidia_config_raw)
+    tolerations = set()
+
+    # Get any tolerations from the plugin configuration
+    if nvidia_config.get("spec", {}) and nvidia_config["spec"].get("template", {}) and nvidia_config["spec"]["template"].get("spec", {}):
+        tolerations.update(Toleration.from_dict(nvidia_config["spec"]["template"]["spec"]["tolerations"]))
```
Comment (amandineslx) [inline python-lib/dku_kube/nvidia_utils.py:30]:
```suggestion
        tolerations.update(Toleration.from_dict(nvidia_config["spec"]["template"]["spec"].get("tolerations", {}))
```

File: python-lib/dku_kube/nvidia_utils.py [hunk 4]
```diff
@@ -10,11 +14,59 @@ def has_gpu_driver(kube_config_path):
...
+        if not _is_none_or_blank(tolerations_json):
+            tolerations.update(Toleration.from_json(tolerations_json))
+
+    # If there are any taints to patch the daemonset with in the node group(s) to create,
+    # we add them to the GPU plugin configuration before updating with another `kubectl apply`
+    tolerations.update(Toleration.from_dict(taints))
+
+    # Patch the Nvidia driver configuration with the tolerations derived from node group(s) taints,
+    # initial Nvidia driver configuration tolerations and Nvidia daemonset tolerations (when applicable)
+    nvidia_config["spec"]["template"]["spec"]["tolerations"] = Toleration.to_list(tolerations)
```
Comment (amandineslx) [inline python-lib/dku_kube/nvidia_utils.py:55]: you have risks of key error here

**Review by amandineslx — CHANGES_REQUESTED** [review]
I did not test, it is only about the code

File: python-lib/dku_utils/taints.py (hunk 2 above)
Comment (vrutz) [reply python-lib/dku_utils/taints.py:7]: We dropped python 2 support for AKS recently. I try to add back compatibility for existing code envs that might not be supported, but I think if you update the plugin, it'll ask you to create a new code env.

File: python-lib/dku_kube/nvidia_utils.py [hunk 5]
```diff
@@ -10,11 +14,70 @@ def has_gpu_driver(kube_config_path):
...
+    nvidia_config = yaml.safe_load(nvidia_config_raw)
+    tolerations = set()
+
+    # Get any tolerations from the plugin configuration
+    if (
+        nvidia_config.get("spec", {})
+        and nvidia_config["spec"].get("template", {})
+        and nvidia_config["spec"]["template"].get("spec", {})
+        and nvidia_config["spec"]["template"]["spec"].get("tolerations", [])
+    ):
```
Comment (amandineslx) [inline python-lib/dku_kube/nvidia_utils.py:34]:
```suggestion
    if (
        "spec" in nvidia_config
        and "template" in nvidia_config["spec"]
        and "spec" in nvidia_config["spec"]["template"]
        and "tolerations" in nvidia_config["spec"]["template"]["spec"]
    ):
```

File: python-lib/dku_kube/nvidia_utils.py [hunk 6]
```diff
@@ -10,11 +14,70 @@ def has_gpu_driver(kube_config_path):
...
+
+    # Patch the Nvidia driver configuration with the tolerations derived from node group(s) taints,
+    # initial Nvidia driver configuration tolerations and Nvidia daemonset tolerations (when applicable)
+    if not nvidia_config.get("spec", None):
+        nvidia_config["spec"] = {}
+    if not nvidia_config["spec"].get("template", None):
+        nvidia_config["spec"]["template"] = {}
+    if not nvidia_config["spec"]["template"].get("spec", None):
+        nvidia_config["spec"]["template"]["spec"] = {}
+    nvidia_config["spec"]["template"]["spec"]["tolerations"] = Toleration.to_list(tolerations)
```
Comment (amandineslx) [inline python-lib/dku_kube/nvidia_utils.py:66]:
```suggestion
    if "spec" not in nvidia_config:
        nvidia_config["spec"] = {}
    if "template" not in nvidia_config["spec"]:
        nvidia_config["spec"]["template"] = {}
    if "spec" not in nvidia_config["spec"]["template"]:
        nvidia_config["spec"]["template"]["spec"] = {}
    nvidia_config["spec"]["template"]["spec"]["tolerations"] = Toleration.to_list(tolerations)
```

**Review by amandineslx — CHANGES_REQUESTED** [review]
code only again, I am deploying a DSS to test

**Review by amandineslx — APPROVED** [review]
OK for me

### Tested
* deploy DSS in elastic fleet via FM
* upgrade plugin with version from PR (commit `9d9590ca2a13db5af3ff442259cbe6b28b9ffd3d`)
* create AKS cluster
  * 1st node pool
    * `Standard_B8ms`
    * 1 node
    * no autoscaling
    * no GPU
  * 2nd node pool
    * `Standard_B8ms`
    * 1 node
    * autoscaling
    * GPU
    * taints: key1=value1:NoSchedule
  * Service CIDR: 10.1.0.0/16
  * DNS IP: 10.1.0.10
* check pods
  * :white_check_mark: one nvidia pod on 1st node pool
  * :white_check_mark: one nvidia pod on 2nd node pool

**Pull request merged**